spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala

```diff
@@ -115,8 +119,61 @@ class DeltaHistoryManager(
           }
         }.map(DeltaHistory.fromCommitInfo)
       }
-    // Spark should return the commits in increasing order as well
-    monotonizeCommitTimestamps(info.collect()).reverse
+    val monotonizedCommits = if (useInCommitTimestamps) {
+      // ICT timestamps are guaranteed to be monotonically increasing.
+      info.collect()
+    } else {
+      monotonizeCommitTimestamps(info.collect())
+    }
+    monotonizedCommits.reverse
+  }
+
+  /**
+   * Get the commit information of the Delta table from commit `[start, end]` in reverse
+   * chronological order. If `end` is `None`, we return all commits from start to now.
+   * @param start The start of the commit range, inclusive.
+   * @param end The end of the commit range, inclusive.
+   */
+  def getHistory(
+      start: Long,
+      end: Option[Long] = None): Seq[DeltaHistory] = {
+    val currentSnapshot = deltaLog.unsafeVolatileSnapshot
+    val (snapshot, resolvedEnd) = end match {
+        case Some(endInclusive) if currentSnapshot.version >= endInclusive =>
+          // Use the cache snapshot if it's fresh enough for the [start, endInclusive] query.
+          (currentSnapshot, math.min(currentSnapshot.version, endInclusive))
+        case _ =>
+          // Either end doesn't exist or the currently cached snapshot isn't new enough to
+          // satisfy it.
+          val snapshot = deltaLog.update()
+          val endInclusive = end.getOrElse(snapshot.version).min(snapshot.version)
+          (snapshot, endInclusive)
+      }
+
+    val commitFileProvider = DeltaCommitFileProvider(snapshot)
+    if (!DeltaConfigs.IN_COMMIT_TIMESTAMPS_ENABLED.fromMetaData(snapshot.metadata)) {
+      getHistoryImpl(start, resolvedEnd, useInCommitTimestamps = false, commitFileProvider)
+    } else {
+      InCommitTimestampUtils.getValidatedICTEnablementInfo(snapshot.metadata) match {
+        case Some(Commit(ictEarliest, _)) =>
+          // getHistoryImpl will return an empty Seq if start > end.
+          val nonICTCommits = getHistoryImpl(
+              start,
+              math.min(resolvedEnd, ictEarliest - 1),
+              useInCommitTimestamps = false,
+              commitFileProvider)
+          val ictCommits = getHistoryImpl(
+              math.max(ictEarliest, start),
+              resolvedEnd,
+              useInCommitTimestamps = true,
+              commitFileProvider)
+          // Merge the two sequences, ensuring ICT commits are listed first as they are more recent,
+          // followed by non-ICT commits, maintaining the reverse chronological order.
+          ictCommits ++ nonICTCommits
```

Comment:
I think it might be a problem that we don't monotonize timestamps across ICT and non-ICT commits?

Reply:
This will only happen for tables where the file modification timestamps of non-ICT commits have been changed after the enablement of ICT (since the first ICT > last non-ICT file modification time). I think this is an acceptable tradeoff. Eventually, these non-ICT commits will expire. It is preferable to ensure that ICT timestamps are unchanging.